File: backend/models/purchase_request.go
package models

import (
	"time"

	"gorm.io/gorm"
)

type PurchaseRequest struct {
	ID           uint      `json:"id" gorm:"primaryKey"`
	Code         string    `json:"code" gorm:"unique;not null;size:20"`
	ProjectID    uint      `json:"project_id" gorm:"not null;index"`
	RequestDate  time.Time `json:"request_date" gorm:"not null"`
	RequiredDate time.Time `json:"required_date"`
	VendorID     *uint     `json:"vendor_id" gorm:"index"` // Optional suggested vendor
	Notes        string    `json:"notes" gorm:"type:text"`
	Status       string    `json:"status" gorm:"size:20;default:'PENDING'"` // PENDING, APPROVED, REJECTED, REVISION, PO_CREATED
	TotalAmount  float64   `json:"total_amount" gorm:"type:decimal(15,2);default:0"`

	// Approval fields
	ApprovedBy      *uint      `json:"approved_by" gorm:"index"`
	ApprovedAt      *time.Time `json:"approved_at"`
	RejectionReason string     `json:"rejection_reason" gorm:"type:text"`

	// CBS Verification fields
	VerifiedBy        *uint      `json:"verified_by" gorm:"index"`
	VerifiedAt        *time.Time `json:"verified_at"`
	VerificationNotes *string    `json:"verification_notes" gorm:"type:text"`

	CreatedBy uint           `json:"created_by" gorm:"not null;index"`
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `json:"-" gorm:"index"`

	// Relations
	Project   Project               `json:"project" gorm:"foreignKey:ProjectID"`
	Requester User                  `json:"requester" gorm:"foreignKey:CreatedBy"`
	Approver  *User                 `json:"approver,omitempty" gorm:"foreignKey:ApprovedBy"`
	Verifier  *User                 `json:"verifier,omitempty" gorm:"foreignKey:VerifiedBy"`
	Vendor    *Vendor               `json:"vendor,omitempty" gorm:"foreignKey:VendorID"`
	Items     []PurchaseRequestItem `json:"items" gorm:"foreignKey:PurchaseRequestID"`
}

type PurchaseRequestItem struct {
	ID                uint    `json:"id" gorm:"primaryKey"`
	PurchaseRequestID uint    `json:"purchase_request_id" gorm:"not null;index"`
	ItemName          string  `json:"item_name" gorm:"not null;size:255"` // Can be free text or product name
	ProductID         *uint   `json:"product_id" gorm:"index"`            // Optional link to product catalog (legacy)
	MaterialID        *uint   `json:"material_id" gorm:"index"`           // Link to master material
	Quantity          float64 `json:"quantity" gorm:"type:decimal(10,2);not null"`
	Unit              string  `json:"unit" gorm:"size:50"`
	EstimatedPrice    float64 `json:"estimated_price" gorm:"type:decimal(15,2);default:0"`
	TotalPrice        float64 `json:"total_price" gorm:"type:decimal(15,2);default:0"`
	Notes             string  `json:"notes" gorm:"type:text"`

	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `json:"-" gorm:"index"`

	// Relations
	PurchaseRequest PurchaseRequest `json:"purchase_request" gorm:"foreignKey:PurchaseRequestID"`
	Material        *Material       `json:"material,omitempty" gorm:"foreignKey:MaterialID"`
}

// Purchase Request Status Constants
const (
	PRStatusPending             = "PENDING"
	PRStatusPendingVerification = "PENDING_VERIFICATION"
	PRStatusVerified            = "VERIFIED"
	PRStatusApproved            = "APPROVED"
	PRStatusRejected            = "REJECTED"
	PRStatusRevision            = "REVISION"
	PRStatusPOCreated           = "PO_CREATED"
	PRStatusCompleted           = "COMPLETED"
)

// MaterialImpact represents the estimated impact of a PR on material stock
type MaterialImpact struct {
	ProductID      *uint   `json:"product_id"`
	ProductName    string  `json:"product_name"`
	ProductCode    string  `json:"product_code"`
	Unit           string  `json:"unit"`
	RequestedQty   float64 `json:"requested_qty"`
	CurrentStock   float64 `json:"current_stock"`
	ProjectedStock float64 `json:"projected_stock"`
	Status         string  `json:"status"` // OK, LOW, CRITICAL
}

File: frontend/src/types/purchaseRequest.ts
export interface PurchaseRequestItem {
    id?: number;
    purchase_request_id?: number;
    item_name: string;
    product_id?: number;
    quantity: number;
    unit: string;
    estimated_price: number;
    total_price: number;
    notes: string;
}

export interface PurchaseRequest {
    id: number;
    code: string;
    project_id: number;
    request_date: string;
    required_date: string;
    vendor_id?: number;
    notes: string;
    status: 'PENDING' | 'APPROVED' | 'REJECTED' | 'REVISION' | 'PO_CREATED';
    total_amount: number;

    // Approval fields
    approved_by?: number;
    approved_at?: string;
    rejection_reason?: string;

    created_by: number;
    created_at: string;
    updated_at: string;

    // Relations
    project?: {
        id: number;
        project_name: string;
    };
    vendor?: {
        id: number;
        name: string;
    };
    requester?: {
        id: number;
        username: string;
        first_name: string;
        last_name: string;
        role: string;
    };
    approver?: {
        id: number;
        name: string;
    };
    items?: PurchaseRequestItem[];
}

export interface CreatePRData {
    project_id: number;
    request_date: string;
    required_date: string;
    vendor_id?: number;
    notes: string;
    items: PurchaseRequestItem[];
}

export interface UpdatePRData extends CreatePRData { }

export interface PRFilter {
    project_id?: number;
    status?: string;
}

export interface MaterialImpact {
    product_id?: number;
    product_name: string;
    product_code: string;
    unit: string;
    requested_qty: number;
    current_stock: number;
    projected_stock: number;
    status: 'OK' | 'LOW' | 'CRITICAL';
}
